src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs

```diff
@@ -63,77 +73,61 @@ public int Count
-            if (cacheEntryOptions.AbsoluteExpirationRelativeToNow.HasValue)
+            if (entry._absoluteExpirationRelativeToNow.HasValue)
             {
-                absoluteExpiration = utcNow + cacheEntryOptions.AbsoluteExpirationRelativeToNow;
+                absoluteExpiration = utcNow + entry._absoluteExpirationRelativeToNow;
             }
-            else if (cacheEntryOptions.AbsoluteExpiration.HasValue)
+            else if (entry._absoluteExpiration.HasValue)
             {
-                if (cacheEntryOptions.AbsoluteExpiration <= utcNow)
-                {
-                    throw new ArgumentOutOfRangeException(
-                        nameof(MemoryCacheEntryOptions.AbsoluteExpiration),
-                        cacheEntryOptions.AbsoluteExpiration.Value,
-                        "The absolute expiration value must be in the future.");
-                }
-
-                absoluteExpiration = cacheEntryOptions.AbsoluteExpiration;
+                absoluteExpiration = entry._absoluteExpiration;
             }
 
-            var entry = new CacheEntry(
-                key,
-                value,
-                utcNow,
-                absoluteExpiration,
-                cacheEntryOptions,
-                _entryExpirationNotification);
-
-            var link = EntryLinkHelpers.ContextLink;
-            if (link != null)
-            {
-                // Copy expiration tokens and AbsoluteExpiration to the link.
-                // We do this regardless of it gets cached because the tokens are associated with the value we'll return.
-                if (entry.Options.ExpirationTokens != null)
-                {
-                    link.AddExpirationTokens(entry.Options.ExpirationTokens);
-                }
-                if (absoluteExpiration.HasValue)
+            // Applying the option's absolute expiration only if it's not already smaller.
+            // This can be the case if a dependent cache entry has a smaller value, and
+            // it was set by cascading it to its parent.
+            if (absoluteExpiration.HasValue)
+            {
+                if (!entry._absoluteExpiration.HasValue || absoluteExpiration.Value < entry._absoluteExpiration.Value)
                 {
-                    link.SetAbsoluteExpiration(absoluteExpiration.Value);
+                    entry._absoluteExpiration = absoluteExpiration;
                 }
             }
 
-            bool added = false;
+            // Initialize the last access timestamp at the time the entry is added
+            entry.LastAccessed = utcNow;
+
+            var added = false;
+            CacheEntry priorEntry;
 
             _entryLock.EnterWriteLock();
             try
             {
-                if (_entries.TryGetValue(key, out priorEntry))
+                
+                if (_entries.TryGetValue(entry.Key, out priorEntry))
                 {
-                    _entries.Remove(key);
```

Comment:
Keep this remove. The new entry may already have expired and fail to overwrite, causing SetExpired to be fired again next time.